Userland/Demos/CatDog/CatDog.cpp

```diff
@@ -10,174 +10,187 @@
 #include <LibGUI/Painter.h>
 #include <LibGUI/Window.h>
 
+namespace {
+
+enum State : uint16_t {
+    ALERT = 0x0000,
+    FRAME_2 = 0x0001,
+    UP = 0x0010,
+    DOWN = 0x0020,
+    LEFT = 0x0040,
+    RIGHT = 0x0080,
+    SLEEPING = 0x0100,
+    IDLE = 0x0200,
+    INSPECTOR = 0x1000,
+    ARTIST = 0x2000,
+};
+
+static constexpr uint16_t DIRECTION_MASK = 0x00F0;
+static constexpr uint16_t SPECIAL_IDLE_MASK = 0xF000;
+
+}
+
 ErrorOr<NonnullRefPtr<CatDog>> CatDog::create()
 {
+    struct ImageSource {
+        uint16_t state;
+        StringView path;
+    };
+
+    // NOTE: The order of the elements is important. Matching is done in best-match order.
+    //       So items with the more bits should be placed before items with less bits to
+    //       ensure correct matching order. This also means that "FRAME_2" has to be first.
+    static constexpr Array<ImageSource, 22> const image_sources = {
```

Comment:
Nit: I think that we can omit the size due to a deduction guide on AK::Array. If that doesn't compile then we totally should make sure that there is a proper CTAD deduction guide in a follow-up later down the line.

Reply:
Size deduction did not work.
Leaving this for someone else to implement CTAD.